File: backend/models/media.py
"""
Phase 12 — Media Generation models.

A `MediaJob` is one stage of work (a single provider call — TTS, avatar render,
clip gen, master, lip-sync, …). A pipeline (talking-head / cinematic / audio) is
an ordered chain of jobs linked by `parent_job_id`; each finished stage produces
a `MediaAsset` that feeds the next. Renders are long-running, so jobs are async:
`queued → processing → done|failed`, advanced by webhook or the poll worker.

See docs/explore-media-generation.md + docs/explore-audio-operations.md.
"""

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.sql import func

from backend.database import Base

# Job lifecycle states.
JOB_STATUSES = (
    "queued",  # ready to submit to its provider
    "processing",  # submitted; awaiting async completion (poll/webhook)
    "awaiting_dependency",  # a later pipeline stage; unblocks when its parent finishes
    "done",
    "failed",
    "canceled",
)


class MediaJob(Base):
    """One provider call in a media pipeline (or a standalone single op)."""

    __tablename__ = "media_jobs"
    __table_args__ = ({"extend_existing": True},)

    id = Column(String, primary_key=True)
    user_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    client_id = Column(String, ForeignKey("clients.id"), nullable=True, index=True)
    project_id = Column(String, ForeignKey("projects.id"), nullable=True, index=True)

    # Pipeline this stage belongs to (talking_head | cinematic | audio_only | single).
    pipeline = Column(String, nullable=True)
    stage_index = Column(Integer, nullable=False, default=0)
    # Self-referential chain (plain String, not FK, to avoid create-order cycles).
    parent_job_id = Column(String, nullable=True, index=True)

    kind = Column(String, nullable=False)  # MediaKind value
    provider = Column(String, nullable=False)  # provider name (stub in dry-run)
    status = Column(String, nullable=False, default="queued", index=True)
    external_id = Column(String, nullable=True, index=True)  # provider job id (async)

    input_json = Column(Text, nullable=True)  # JSON spec for the provider
    output_asset_id = Column(String, nullable=True)  # MediaAsset produced (plain str)
    cost_cents = Column(Integer, nullable=False, default=0)
    error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, nullable=False, default=0)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())


class MediaAsset(Base):
    """A produced audio/video artifact (a stage output or the final deliverable)."""

    __tablename__ = "media_assets"
    __table_args__ = ({"extend_existing": True},)

    id = Column(String, primary_key=True)
    user_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    job_id = Column(String, ForeignKey("media_jobs.id"), nullable=False, index=True)
    kind = Column(String, nullable=False)  # audio | clip | video | final
    url = Column(String, nullable=False)  # location (stub URL in dry-run)
    duration_s = Column(Integer, nullable=True)
    mime = Column(String, nullable=True)
    bytes = Column(Integer, nullable=True)
    content_hash = Column(String, nullable=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

File: operator-dashboard/src/api/media.ts
import apiClient from './client';

// Phase 12 — Media generation (audio + video).

export interface MediaAsset {
  id: string;
  kind: string; // audio | clip | video | final
  url: string; // durable storage key (sign via assetUrl())
  duration_s?: number | null;
  mime?: string | null;
}

export interface MediaJob {
  id: string;
  pipeline?: string | null;
  pipeline_run_id?: string | null;
  stage_index: number;
  kind: string;
  provider: string;
  status: string; // queued | processing | awaiting_dependency | done | failed | canceled
  external_id?: string | null;
  output_asset_id?: string | null;
  cost_cents: number;
  error_message?: string | null;
  retry_count: number;
}

export interface MediaJobDetail extends MediaJob {
  assets: MediaAsset[];
}

export interface StageEstimate {
  kind: string;
  provider: string;
  cost_cents: number;
}
export interface Estimate {
  pipeline: string;
  stages: StageEstimate[];
  total_cost_cents: number;
}

export interface GenerateResponse {
  estimate: Estimate;
  confirmed: boolean;
  root_job?: MediaJob;
}

export interface GenerateBody {
  pipeline?: string;
  kind?: string;
  spec: Record<string, unknown>;
  client_id?: string;
  project_id?: string;
  confirm: boolean;
}

export interface PipelineShape {
  [name: string]: { kind: string; provider: string }[];
}

export const mediaApi = {
  async pipelines(): Promise<PipelineShape> {
    const { data } = await apiClient.get<{ pipelines: PipelineShape }>('/api/media/pipelines');
    return data.pipelines;
  },
  async generate(body: GenerateBody): Promise<GenerateResponse> {
    const { data } = await apiClient.post<GenerateResponse>('/api/media/generate', body);
    return data;
  },
  async jobs(params?: { status?: string; pipeline?: string; run_id?: string }): Promise<MediaJob[]> {
    const { data } = await apiClient.get<MediaJob[]>('/api/media/jobs', { params });
    return data;
  },
  async job(id: string): Promise<MediaJobDetail> {
    const { data } = await apiClient.get<MediaJobDetail>(`/api/media/jobs/${id}`);
    return data;
  },
  async cancelJob(id: string): Promise<MediaJob> {
    const { data } = await apiClient.post<MediaJob>(`/api/media/jobs/${id}/cancel`);
    return data;
  },
  async cancelRun(runId: string): Promise<{ run_id: string; canceled: number }> {
    const { data } = await apiClient.post(`/api/media/runs/${runId}/cancel`);
    return data;
  },
  // Fresh signed URL for an asset (JSON — usable from the browser without the 302).
  async assetUrl(assetId: string): Promise<string> {
    const { data } = await apiClient.get<{ url: string }>(`/api/media/assets/${assetId}/url`);
    return data.url;
  },
};
